File: java/leetcode/P1371.java
public class P1371 {

	public int findTheLongestSubstring(String s) {
		int maxlength = 0;
		int[] letters = new int[26];
		letters['a' - 'a'] = 1;
		letters['e' - 'a'] = 2;
		letters['i' - 'a'] = 4;
		letters['o' - 'a'] = 8;
		letters['u' - 'a'] = 16;

		int[] states = new int[32];
		for (int i = 0; i < states.length; i++) {
			states[i] = -1;
		}

		int xor = 0;

		for (int i = 0; i < s.length(); i++) {
			xor ^= letters[s.charAt(i) - 'a'];
			if (xor != 0 && states[xor] == -1) {
				states[xor] = i;
			}
			int tempLength = i - states[xor];
			maxlength = maxlength > tempLength ? maxlength : tempLength;
		}

		return maxlength;
	}

}

File: python/leetcode/P1371.py
class Solution:
	def findTheLongestSubstring(self, s: str) -> int:
		maxlength = 0
		letters = []
		for i in range(26):
			if i is ord('a') - ord('a'):
				letters.append(1)
			elif i is ord('e') - ord('a'):
				letters.append(2)
			elif i is ord('i') - ord('a'):
				letters.append(4)
			elif i is ord('o') - ord('a'):
				letters.append(8)
			elif i is ord('u') - ord('a'):
				letters.append(16)
			else:
				letters.append(0)
		
		states = []
		for i in range(32):
			states.append(-1)

		xor = 0
		for idx, letter in enumerate(s):
			xor ^= letters[ord(letter) - ord('a')]
			if not xor == 0 and states[xor] == -1:
				states[xor] = idx
			maxlength = max(maxlength, (idx - states[xor]))
		
		return maxlength
